samples/nrf9160/serial_lte_modem/src/slm_at_mqtt.c

```diff
@@ -0,0 +1,824 @@
+static struct sockaddr_storage broker;
+
+/* Connected flag */
+static K_SEM_DEFINE(mqtt_connected, 0, 1);
+
+/* File descriptor */
+static struct pollfd fds = {
+	.fd = INVALID_FDS,
+};
+
+/**@brief Function to read the published payload.
+ */
+static int publish_get_payload(struct mqtt_client *c, size_t length)
+{
+	if (length > sizeof(payload_buf)) {
+		return -EMSGSIZE;
+	}
+
+	return mqtt_readall_publish_payload(c, payload_buf, length);
+}
+
+/**@brief Function to handle received publish event.
+ */
+static void handle_mqtt_publish_evt(struct mqtt_client *const c,
+					const struct mqtt_evt *evt)
+{
+	int ret;
+
+	ret = publish_get_payload(c,
+				evt->param.publish.message.payload.len);
+	if (ret < 0) {
+		LOG_ERR("mqtt_read_publish_payload: Failed! %d\n", ret);
+		sprintf(rsp_buf, "#XMQTTMSG: %d,0,0\r\n", ret);
+		rsp_send(rsp_buf, strlen(rsp_buf));
+		return;
+	}
+
+	ret = slm_util_hex_check(payload_buf,
+				evt->param.publish.message.payload.len);
+	if (ret) {
+		char *data_hex;
+		int size;
+
+		size = evt->param.publish.message.payload.len * 2;
+		data_hex = k_malloc(size);
+		if (!data_hex) {
+			LOG_ERR("mqtt_read_publish_payload: Failed! %d\n",
+				-ENOMEM);
+			sprintf(rsp_buf, "#XMQTTMSG: %d,0,0\r\n", -ENOMEM);
+			rsp_send(rsp_buf, strlen(rsp_buf));
+			return;
+		}
+		ret = slm_util_htoa((const u8_t *)&payload_buf,
+				evt->param.publish.message.payload.len,
+				data_hex, size);
+		if (ret < 0) {
+			k_free(data_hex);
+			sprintf(rsp_buf, "#XMQTTMSG: %d,0,0\r\n", ret);
+			rsp_send(rsp_buf, strlen(rsp_buf));
+			return;
+		}
+		sprintf(rsp_buf, "#XMQTTMSG: %d,%d,%d\r\n",
+			DATATYPE_HEXADECIMAL,
+			evt->param.publish.message.topic.topic.size,
+			ret);
+		rsp_send(rsp_buf, strlen(rsp_buf));
+		rsp_send(evt->param.publish.message.topic.topic.utf8,
+			evt->param.publish.message.topic.topic.size);
+		rsp_send("\r\n", 2);
+		rsp_send(data_hex, ret);
+		rsp_send("\r\n", 2);
+		k_free(data_hex);
+	} else {
+		sprintf(rsp_buf, "#XMQTTMSG: %d,%d,%d\r\n",
+			DATATYPE_PLAINTEXT,
+			evt->param.publish.message.topic.topic.size,
+			evt->param.publish.message.payload.len);
+		rsp_send(rsp_buf, strlen(rsp_buf));
+		rsp_send(evt->param.publish.message.topic.topic.utf8,
+			evt->param.publish.message.topic.topic.size);
+		rsp_send("\r\n", 2);
+		rsp_send(payload_buf,
+			evt->param.publish.message.payload.len);
+		rsp_send("\r\n", 2);
+	}
+}
+
+/**@brief MQTT client event handler
+ */
+void mqtt_evt_handler(struct mqtt_client *const c,
+		      const struct mqtt_evt *evt)
+{
+	int ret;
+
+	sprintf(rsp_buf, "#XMQTTEVT: %d,%d\r\n",
+		evt->type, evt->result);
+	rsp_send(rsp_buf, strlen(rsp_buf));
+
+	switch (evt->type) {
+	case MQTT_EVT_CONNACK:
+		if (evt->result != 0) {
+			LOG_ERR("MQTT connect failed %d", evt->result);
+			ctx.connected = false;
+			break;
+		}
+		LOG_DBG("MQTT client connected!");
+		break;
+
+	case MQTT_EVT_DISCONNECT:
+		if (evt->result != 0) {
+			LOG_ERR("MQTT disconnect failed %d", evt->result);
+			break;
```

Comment:
could you rephrase the code to avoid two breaks for one case?

Reply:
Rephrased.